controllers/configuration/backend/backend.go

```diff
@@ -0,0 +1,316 @@
+	}
+
+	templateVars := backendVars{
+		SecretSuffix:    backend.SecretSuffix,
+		InClusterConfig: backend.InClusterConfig,
+		Namespace:       namespace,
+	}
+	var wr bytes.Buffer
+	err = tmpl.Execute(&wr, templateVars)
+	if err != nil {
+		return "", err
+	}
+	return wr.String(), nil
+}
+
+// ParseConfigurationBackend parses backend Conf from the v1beta2.Configuration
+func ParseConfigurationBackend(configuration *v1beta2.Configuration, terraformBackendNamespace string) (*Conf, error) {
+	backend := configuration.Spec.Backend
+
+	switch {
+
+	case backend != nil && len(backend.Inline) > 0:
+		// In this case, use the inline custom backend
+		backendTF, backendType, err := handleInlineBackendHCL(backend.Inline)
+		if err != nil {
+			return nil, err
+		}
+		return &Conf{
+			BackendType: backendType,
+			HCL:         backendTF,
+			UseCustom:   true,
+		}, nil
+
+	case backend != nil && len(backend.BackendType) > 0:
+		// In this case, use the explicit custom backend
+
+		// first, check if is valid custom backend
+		backendType := backend.BackendType
+		// fetch backendConfValue using reflection
+		backendStructValue := reflect.ValueOf(backend)
+		if backendStructValue.Kind() == reflect.Ptr {
+			backendStructValue = backendStructValue.Elem()
+		}
+		backendField := backendStructValue.FieldByNameFunc(func(name string) bool {
+			return strings.EqualFold(name, string(backendType))
+		})
+		if backendField.IsNil() {
+			return nil, fmt.Errorf("there is no configuration for backendType %s", backend.BackendType)
+		}
+		backendConfValue := backendField.Interface()
+
+		// second, handle the backendConf
+		backendHCL, backendType, secretList := handleExplicitBackend(backendConfValue, backendType)
+		return &Conf{
+			BackendType: backendType,
+			HCL:         backendHCL,
+			UseCustom:   true,
+			Secrets:     secretList,
+		}, nil
+
+	case backend != nil && len(backend.BackendType) == 0:
+		// In this case, use the default k8s backend
+		klog.Warningf("the spec.backend.backendType of Configuration{Namespace: %s, Name: %s} is empty, use the default kubernetes backend", configuration.Namespace, configuration.Name)
+		fallthrough // down to default
+
+	default:
+		// use the default k8s backend
+		if configuration.Spec.Backend != nil {
+			if configuration.Spec.Backend.SecretSuffix == "" {
+				configuration.Spec.Backend.SecretSuffix = configuration.Name
+			}
+			configuration.Spec.Backend.InClusterConfig = true
+		} else {
+			configuration.Spec.Backend = &v1beta2.Backend{
+				SecretSuffix:    configuration.Name,
+				InClusterConfig: true,
+			}
+		}
+		backendTF, err := renderTemplate(configuration.Spec.Backend, terraformBackendNamespace)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to prepare Terraform backend configuration")
+		}
+		return &Conf{
+			BackendType: "kubernetes",
+			HCL:         backendTF,
+			UseCustom:   false,
+		}, nil
+	}
+
+}
+
+func handleInlineBackendHCL(code string) (string, v1beta2.BackendType, error) {
```

Comment:
Why do you handle an inlined backend? How about just treating it as `state.tf`?

Reply:
1. we need to get the backendType from the inlined backend and the backendType will be useful when fetch the Terraform state.
2. We should check if there are invalid fields in the inlined backend (such some fileds related to the local file system).